aten/src/ATen/native/cuda/Nonzero.cu

```diff
@@ -183,6 +214,84 @@ void nonzero_cuda_out_impl(const Tensor& self, Tensor& out) {
   }
 }
 
+template <typename scalar_t>
+void nonzero_static_cuda_out_impl(
+    const Tensor& self,
+    int64_t size,
+    int64_t fill_value,
+    Tensor& out) {
+# if (defined(CUDA_VERSION) && CUDA_VERSION > 11040) || defined(USE_ROCM)
+
+  Tensor self_ = self.contiguous();
+  // see comment in nonzero_cuda_out_impl on reqs for out
+  bool out_correct_size =
+      out.dim() == 2 && out.sizes()[0] == size && out.sizes()[1] == self.dim();
+  bool need_to_copy = out_correct_size && !out.t().is_contiguous();
+  if (!out_correct_size) {
+    out.resize_({self.dim(), size}).t();
+  }
+  if (out.numel() == 0) return;
+  // we need to allocate temporary out to then copy to user provided out
+  at::Tensor out_temp;
+  if (need_to_copy) {
+    out_temp =
+        Tensor(at::detail::empty_cuda({self.dim(), size}, out.options())).t();
+  }
+  int64_t* out_data_ptr = need_to_copy ? out_temp.mutable_data_ptr<int64_t>()
+                                       : out.mutable_data_ptr<int64_t>();
+
+  const scalar_t * in_data_ptr = self_.const_data_ptr<scalar_t>();
+  constexpr int BLOCK_THREADS = 512; //block_threads<sizeof(scalar_t)>();
+  constexpr int ITEMS_PER_THREAD = 16;
+  auto grid_size = (self.numel() + BLOCK_THREADS * ITEMS_PER_THREAD - 1) / (BLOCK_THREADS * ITEMS_PER_THREAD);
+  const int64_t num_sms = at::cuda::getCurrentDeviceProperties()->multiProcessorCount;
+
+  const int iters_per_cta = (grid_size + num_sms - 1)/num_sms;
+  grid_size = (self.numel() + iters_per_cta * BLOCK_THREADS * ITEMS_PER_THREAD - 1) / (iters_per_cta * BLOCK_THREADS * ITEMS_PER_THREAD);
+  //grid_size = std::min(num_sms, grid_size);
+  auto& allocator = *c10::cuda::CUDACachingAllocator::get();
+  auto agg = allocator.allocate(grid_size * sizeof(int));
+  at::cuda::cub::calc_block_sums<BLOCK_THREADS, ITEMS_PER_THREAD, true>
```

Comment:
I see that the non-static cuda implementation was recently changed to handle large tensors with chunking. Does this code have to adapt similarly? Is this handled by computing how many iterations each CTA has to go through?

Reply:
In this case we are splitting a tensor between threadblocks, so even if the tensor is large each threadblock is handling only a chunk that's < INT_MAX, so there's no additional chunking needed.